File: ui/src/api/offline.ts
// ── Offline Queue & Cloud Sync ────────────────────────────────────

import { loggedInvoke } from '@/utils/logged-invoke';

// ── Offline Queue ────────────────────────────────────────────────

/** An item in the offline action queue. */
export interface OfflineQueueItemDto {
  id: string;
  action: string;
  /**
   * JSON-serialized action payload (SYNC-11: present in the Rust
   * `OfflineQueueItemDto` serializer — must not be dropped here).
   */
  payload: string;
  status: string;
  retryCount: number;
  lastError: string | null;
  createdAt: string;
  syncedAt: string | null;
  /** Tenant / store ID for multi-store isolation (OFF-09). */
  tenantId: string;
  /** Sync priority tier: "critical" | "normal" | "low" (OFF-09). */
  priority: string;
}

/** Arguments for enqueuing an offline action. */
export interface EnqueueOfflineArgs {
  action: string;
  payload: string;
  /** Optional tenant / store ID (OFF-09). Defaults to "default". */
  tenantId?: string;
  /** Optional sync priority tier (OFF-09): "critical" | "normal" | "low". */
  priority?: 'critical' | 'normal' | 'low';
}

/**
 * Result of a manual retry of the pending offline queue.
 *
 * SYNC-04 / SYNC-11: matches the Rust `SyncResult` DTO exactly
 * (camelCase `syncedCount` / `failedCount` / `totalCount`).
 */
export interface SyncResult {
  /** Number of items successfully synced. */
  syncedCount: number;
  /** Number of items that failed to sync. */
  failedCount: number;
  /** Total number of items that were attempted. */
  totalCount: number;
  /** The server rejected the attempt because this tenant is on the free
   *  plan (ADR sync-plan-gating) — show an upgrade prompt. */
  planRequired?: boolean;
}

/** Summary of offline queue status (P1-6 sync observability). */
export interface OfflineQueueSummaryDto {
  pendingCount: number;
  syncedCount: number;
  failedCount: number;
  /** Number of items resolved via conflict during sync (P1-3). */
  conflictCount: number;
  lastSyncedAt: string | null;
  oldestPendingAt: string | null;
}

/** Get offline queue status summary (scoped — ADR #7). */
export const getOfflineQueueStatusSummaryScoped = (sessionToken: string): Promise<OfflineQueueSummaryDto> =>
  loggedInvoke<OfflineQueueSummaryDto>('offline_queue_status_summary_scoped', { sessionToken });

// ── Offline queue operations (scoped — ADR #7) ──────────────────────

/** List all offline actions (scoped — ADR #7). */
export const listAllOfflineScoped = (sessionToken: string): Promise<OfflineQueueItemDto[]> =>
  loggedInvoke<OfflineQueueItemDto[]>('list_all_offline_scoped', { sessionToken });

/** Get the count of pending offline actions (scoped — ADR #7). */
export const pendingOfflineCountScoped = (sessionToken: string): Promise<number> =>
  loggedInvoke<number>('pending_offline_count_scoped', { sessionToken });

/** Retry syncing all pending offline actions (scoped — ADR #7). */
export const retryOfflineSyncScoped = (sessionToken: string): Promise<SyncResult> =>
  loggedInvoke<SyncResult>('retry_offline_sync_scoped', { sessionToken });

/** Delete an offline queue item (scoped). */
export const deleteOfflineItemScoped = (sessionToken: string, id: string): Promise<void> =>
  loggedInvoke('delete_offline_item_scoped', { sessionToken, args: { id } });

// ── Remote Dead-Letter (quarantined pulls) ────────────────────────

/**
 * A remote sync item that repeatedly failed to apply during a pull and
 * was quarantined in `sync_remote_failures` (SYNC-09 / SYNC-11).
 *
 * Matches the Rust `RemoteSyncFailureDto` serializer exactly
 * (camelCase `itemId` / `deadLettered`).
 */
export interface RemoteSyncFailureDto {
  /** Remote item identifier. */
  itemId: string;
  /** Remote action name. */
  action: string;
  /** Original payload retained for operator inspection. */
  payload: string;
  /** Number of failed application attempts. */
  attempts: number;
  /** Most recent application error. */
  lastError: string;
  /** Whether retry is exhausted and the item is quarantined. */
  deadLettered: boolean;
}

/** List remote failures (scoped — ADR #7). */
export const listRemoteFailuresScoped = (sessionToken: string): Promise<RemoteSyncFailureDto[]> =>
  loggedInvoke<RemoteSyncFailureDto[]>('list_remote_failures_scoped', { sessionToken });

/** Requeue a dead-lettered remote item (scoped). */
export const requeueRemoteFailureScoped = (sessionToken: string, itemId: string): Promise<void> =>
  loggedInvoke('requeue_remote_failure_scoped', { sessionToken, args: { itemId } });

// ── Cloud Sync Settings ──────────────────────────────────────────

/** Cloud sync configuration. */
export interface SyncSettingsDto {
  serverUrl: string | null;
  hasApiKey: boolean;
  enabled: boolean;
}

/** Arguments for updating cloud sync settings. */
export interface UpdateSyncSettingsArgs {
  serverUrl?: string | null;
  apiKey?: string | null;
  enabled: boolean;
}

/** Result of a sync run, including per-batch success/failure counts. */
export interface SyncAttemptResult {
  synced: number;
  failed: number;
  error: string | null;
  /** The server rejected the attempt because this tenant is on the free
   *  plan (ADR sync-plan-gating) — show an upgrade prompt. */
  planRequired?: boolean;
}

/** Result of pulling data from the cloud server. */
export interface PullResult {
  productsPulled: number;
  taxRatesPulled: number;
  usersPulled: number;
  error: string | null;
}

/** Get cloud sync settings resolved from a session token. ADR #7. */
export const getSyncSettingsScoped = (sessionToken: string): Promise<SyncSettingsDto> =>
  loggedInvoke<SyncSettingsDto>('get_sync_settings_scoped', { sessionToken });

/** Update the cloud sync settings (scoped — ADR #7). */
export const updateSyncSettingsScoped = (sessionToken: string, args: UpdateSyncSettingsArgs): Promise<void> =>
  loggedInvoke<void>('update_sync_settings_scoped', { sessionToken, args });

/** Run a sync cycle (scoped — ADR #7). */
export const syncRunScoped = (sessionToken: string): Promise<SyncAttemptResult> =>
  loggedInvoke<SyncAttemptResult>('sync_run_scoped', { sessionToken });

// ── PostgreSQL sync settings & daemon ──────────────────────────────

/** PostgreSQL sync configuration (the PG transport's connection settings). */
export interface PgSyncSettingsDto {
  enabled: boolean;
  host: string | null;
  port: string | null;
  dbname: string | null;
  user: string | null;
  hasPassword: boolean;
}

/** Arguments for updating PostgreSQL sync settings. `password` is only
 *  written when provided — omit it to keep the stored secret untouched
 *  (the UI sends `undefined` for the unmasked field). */
export interface UpdatePgSyncSettingsArgs {
  enabled: boolean;
  host?: string | null;
  port?: string | null;
  dbname?: string | null;
  user?: string | null;
  password?: string | null;
}

/** Snapshot of the PG sync daemon's state (camelCase mirror of the Rust
 *  `PgDaemonStatus` serializer). */
export interface PgDaemonStatusDto {
  running: boolean;
  lastSyncAt: string | null;
  lastPushed: number;
  lastPulled: number;
  lastError: string | null;
  pendingCount: number;
}

/** Get the PostgreSQL sync settings (scoped — ADR #7). */
export const getPgSyncSettingsScoped = (sessionToken: string): Promise<PgSyncSettingsDto> =>
  loggedInvoke<PgSyncSettingsDto>('get_pg_sync_settings_scoped', { sessionToken });

/** Update the PostgreSQL sync settings (scoped — ADR #7). */
export const updatePgSyncSettingsScoped = (sessionToken: string, args: UpdatePgSyncSettingsArgs): Promise<void> =>
  loggedInvoke<void>('update_pg_sync_settings_scoped', { sessionToken, args });

/** Get the PG sync daemon's current status (scoped — ADR #7). */
export const pgSyncStatusScoped = (sessionToken: string): Promise<PgDaemonStatusDto> =>
  loggedInvoke<PgDaemonStatusDto>('pg_sync_status_scoped', { sessionToken });

/** Start the background PG sync daemon (scoped — ADR #7; no-op when already running). */
export const pgSyncStartScoped = (sessionToken: string): Promise<void> =>
  loggedInvoke<void>('pg_sync_start_scoped', { sessionToken });

/** Stop the background PG sync daemon (scoped — ADR #7; no-op when not running). */
export const pgSyncStopScoped = (sessionToken: string): Promise<void> =>
  loggedInvoke<void>('pg_sync_stop_scoped', { sessionToken });

/**
 * Arguments for a destructive snapshot pull.
 *
 * SYNC-03: `confirmDestructive` must be `true` for the backend command
 * to proceed — it rejects any call without explicit user consent.
 */
export interface SyncPullArgs {
  confirmDestructive: boolean;
}

/**
 * Pull data (products, tax rates, users) from the cloud server.
 *
 * SYNC-03: the destructive consent is part of the IPC contract — the
 * caller must pass `{ confirmDestructive: true }` after showing a
 * confirmation dialog. (scoped — ADR #7)
 */
export const syncPullScoped = (sessionToken: string, args: SyncPullArgs): Promise<PullResult> =>
  loggedInvoke<PullResult>('sync_pull_scoped', { sessionToken, args });

// ── Tenant plan (ADR sync-plan-gating) ─────────────────────────

/** Result of reading the caller's own sync plan from the server. */
export interface SyncPlanResult {
  ok: boolean;
  /** Effective plan: `free` | `pro` — present when the read succeeded. */
  plan: 'free' | 'pro' | null;
  status: string;
}

/** Get sync plan (scoped — ADR #7). */
export const getSyncPlanScoped = (sessionToken: string): Promise<SyncPlanResult> =>
  loggedInvoke<SyncPlanResult>('get_sync_plan_scoped', { sessionToken });

// ── Connection Test ──────────────────────────────────────────────

/** Result of pinging the cloud server's health endpoint. */
export interface PingResult {
  ok: boolean;
  status: string;
  latencyMs: number | null;
}

/** Test connectivity to the configured cloud server (H-6: URL always resolved from saved settings). */
export const testSyncConnection = (): Promise<PingResult> =>
  loggedInvoke<PingResult>('test_sync_connection');

/** Test connectivity (scoped — ADR #7). */
export const testSyncConnectionScoped = (sessionToken: string): Promise<PingResult> =>
  loggedInvoke<PingResult>('test_sync_connection_scoped', { sessionToken });

// ── Token Request ────────────────────────────────────────────────

/** Result of requesting a new JWT API token from the cloud server. */
export interface TokenResult {
  ok: boolean;
  token: string | null;
  status: string;
  expiresAt: string | null;
}

/** Request a new JWT token (scoped — ADR #7). */
export const requestSyncTokenScoped = (sessionToken: string): Promise<TokenResult> =>
  loggedInvoke<TokenResult>('request_sync_token_scoped', { sessionToken });

File: apps/desktop-client/src/commands/offline.rs
//! Offline queue commands.
//!
//! These commands allow the front-end to enqueue, list, and sync
//! transactions that were created while the network was unavailable.

use serde::{Deserialize, Serialize};
use tauri::State;

use oz_core::sync_client::{self, SyncAttemptResult, SyncConfig};
use oz_core::{OfflineQueueItem, RemoteSyncFailure, Store, SyncPriority};

use foundation::validate_not_empty;

use crate::commands::authz::require_permission_for_session;
use crate::error::AppError;
use crate::state::AppState;
use oz_core::permissions;

// ── DTOs ──────────────────────────────────────────────────────────────

/// Offline queue item DTO for the front-end.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OfflineQueueItemDto {
    /// Unique identifier.
    pub id: String,
    /// Action.
    pub action: String,
    /// Payload.
    pub payload: String,
    /// Current status.
    pub status: String,
    /// Retry Count.
    pub retry_count: i64,
    /// Last Error.
    pub last_error: Option<String>,
    /// ISO-8601 creation timestamp.
    pub created_at: String,
    /// Synced At.
    pub synced_at: Option<String>,
    /// Tenant / store ID for multi-store isolation (OFF-09).
    pub tenant_id: String,
    /// Sync priority tier: "critical" | "normal" | "low" (OFF-09).
    pub priority: String,
}

impl From<OfflineQueueItem> for OfflineQueueItemDto {
    fn from(item: OfflineQueueItem) -> Self {
        Self {
            id: item.id,
            action: item.action,
            payload: item.payload,
            status: item.status.as_stored_str().to_owned(),
            retry_count: item.retry_count,
            last_error: item.last_error,
            created_at: item.created_at,
            synced_at: item.synced_at,
            tenant_id: item.tenant_id,
            priority: item.priority.as_str().to_owned(),
        }
    }
}

/// Retained remote-application failure DTO for the front-end.
///
/// Exposes everything an operator needs to decide whether to requeue a
/// dead-lettered item (via `requeue_remote_failure`): the remote item id,
/// action, retained payload for inspection, attempt count, the latest
/// error, and the dead-letter flag.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteSyncFailureDto {
    /// Remote item identifier.
    pub item_id: String,
    /// Remote action name.
    pub action: String,
    /// Original payload retained for operator inspection.
    pub payload: String,
    /// Number of failed application attempts.
    pub attempts: i64,
    /// Most recent application error.
    pub last_error: String,
    /// Whether retry is exhausted and the item is quarantined.
    pub dead_lettered: bool,
}

impl From<RemoteSyncFailure> for RemoteSyncFailureDto {
    fn from(failure: RemoteSyncFailure) -> Self {
        Self {
            item_id: failure.item_id,
            action: failure.action,
            payload: failure.payload,
            attempts: failure.attempts,
            last_error: failure.last_error,
            dead_lettered: failure.dead_lettered,
        }
    }
}

/// Result of a sync retry attempt.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncResult {
    /// Number of items successfully synced.
    pub synced_count: i64,
    /// Number of items that failed to sync.
    pub failed_count: i64,
    /// Total number of items that were attempted.
    pub total_count: i64,
    /// The server rejected the attempt because this tenant is on the
    /// `free` plan (ADR sync-plan-gating). Items stay `pending` and sync
    /// automatically after an upgrade.
    #[serde(default)]
    pub plan_required: bool,
}

/// Arguments for enqueuing an offline transaction.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnqueueOfflineArgs {
    /// The action to perform (e.g. "complete_sale", "void_sale").
    pub action: String,
    /// JSON-serialized payload for the action.
    pub payload: String,
    /// Optional tenant / store ID (OFF-09). Defaults to "default" for
    /// single-store deployments.
    #[serde(default)]
    pub tenant_id: Option<String>,
    /// Optional sync priority tier (OFF-09): "critical" | "normal" | "low".
    #[serde(default)]
    pub priority: Option<String>,
}

// ── Commands ──────────────────────────────────────────────────────────

fn run_list_pending_offline(
    conn: &rusqlite::Connection,
) -> Result<Vec<OfflineQueueItemDto>, AppError> {
    let store = Store::new(conn);
    let items = store.list_pending_offline()?;
    let dtos: Vec<OfflineQueueItemDto> = items.into_iter().map(OfflineQueueItemDto::from).collect();
    Ok(dtos)
}

/// Summary of offline queue status — counts by status and sync timing.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OfflineQueueSummaryDto {
    /// Number of pending (unsynced) items.
    pub pending_count: i64,
    /// Number of successfully synced items.
    pub synced_count: i64,
    /// Number of failed items.
    pub failed_count: i64,
    /// Number of items resolved via conflict (P1-3).
    pub conflict_count: i64,
    /// ISO-8601 timestamp of the most recently synced item, if any.
    pub last_synced_at: Option<String>,
    /// ISO-8601 timestamp of the oldest pending item, if any.
    pub oldest_pending_at: Option<String>,
}

/// Arguments for `requeue_remote_failure`.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RequeueRemoteFailureArgs {
    /// Remote item id currently quarantined in `sync_remote_failures`.
    pub item_id: String,
}

/// Execute the requeue against a connection, extracted so the command
/// boundary is unit-testable without a Tauri runtime.
fn run_requeue_remote_failure(conn: &rusqlite::Connection, item_id: &str) -> Result<(), AppError> {
    let store = Store::new(conn);
    store.requeue_remote_failure(item_id)?;
    Ok(())
}

/// Execute the listing against a connection, extracted so the command
/// boundary is unit-testable without a Tauri runtime.
fn run_list_remote_failures(
    conn: &rusqlite::Connection,
) -> Result<Vec<RemoteSyncFailureDto>, AppError> {
    let store = Store::new(conn);
    let failures = store.list_remote_failures()?;
    Ok(failures
        .into_iter()
        .map(RemoteSyncFailureDto::from)
        .collect())
}

#[cfg(test)]
#[path = "offline_tests.rs"]
mod tests;

// ── Scoped variants (ADR #7) ────────────────────────────────────────

/// Enqueue a transaction for later sync (scoped).
#[tauri::command]
pub async fn enqueue_offline_scoped(
    args: EnqueueOfflineArgs,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<OfflineQueueItemDto, AppError> {
    validate_not_empty("action", &args.action).map_err(|e| AppError::Invalid(e.to_string()))?;
    validate_not_empty("payload", &args.payload).map_err(|e| AppError::Invalid(e.to_string()))?;

    let tenant_id = args.tenant_id.as_deref().unwrap_or("default");
    let priority = args
        .priority
        .as_deref()
        .map(SyncPriority::from_str_lenient)
        .unwrap_or(SyncPriority::Normal);

    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let item = store.enqueue_offline_scoped(&args.action, &args.payload, tenant_id, priority)?;
    drop(db);

    tracing::info!(id = %item.id, action = %item.action, tenant_id, "offline transaction enqueued (scoped)");
    Ok(item.into())
}

/// List all pending (unsynced) offline queue items (scoped).
#[tauri::command]
pub async fn list_pending_offline_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<OfflineQueueItemDto>, AppError> {
    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    run_list_pending_offline(&db)
}

/// List all offline queue items (scoped).
#[tauri::command]
pub async fn list_all_offline_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<OfflineQueueItemDto>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    // F-017: enforce per-domain permission on this scoped command.
    require_permission_for_session(&state, &session, permissions::SYNC_MANAGE).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let items = store.list_all_offline()?;
    let dtos: Vec<OfflineQueueItemDto> = items.into_iter().map(OfflineQueueItemDto::from).collect();
    Ok(dtos)
}

/// Get a summary of the offline queue status (scoped).
#[tauri::command]
pub async fn offline_queue_status_summary_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<OfflineQueueSummaryDto, AppError> {
    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let summary = store.offline_queue_status_summary()?;
    drop(db);
    Ok(OfflineQueueSummaryDto {
        pending_count: summary.pending_count,
        synced_count: summary.synced_count,
        failed_count: summary.failed_count,
        conflict_count: summary.conflict_count,
        last_synced_at: summary.last_synced_at,
        oldest_pending_at: summary.oldest_pending_at,
    })
}

/// Get the count of pending offline items (scoped).
#[tauri::command]
pub async fn pending_offline_count_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<i64, AppError> {
    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let count = store.pending_offline_count()?;
    drop(db);
    Ok(count)
}

/// Attempt to sync all pending offline items (scoped).
#[tauri::command]
pub async fn retry_offline_sync_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<SyncResult, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    // F-017: enforce per-domain permission on this scoped command.
    require_permission_for_session(&state, &session, permissions::SYNC_MANAGE).await?;
    let (pending_items, config_opt) = {
        let db = conn
            .lock()
            .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
        let store = Store::new(&db);
        let pending = store.list_pending_offline()?;
        let config = SyncConfig::from_settings(&store)?;
        (pending, config)
    };

    let total_count = pending_items.len() as i64;
    let config = match config_opt {
        Some(c) => c,
        None => {
            return Err(AppError::Invalid(
                "Sync is not configured or disabled — items remain pending".into(),
            ));
        }
    };

    if pending_items.is_empty() {
        return Ok(SyncResult {
            synced_count: 0,
            failed_count: 0,
            total_count: 0,
            plan_required: false,
        });
    }

    let mut pending_items = pending_items;
    pending_items.sort_by_key(|i| i.priority);

    let outcomes = sync_client::send_items_to_server(&config, &pending_items).await;

    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let attempt = match outcomes {
        Ok(outcomes) => sync_client::apply_sync_outcomes(&store, &pending_items, &outcomes)?,
        Err(sync_client::SyncHttpError::PlanRequired) => SyncAttemptResult {
            synced: 0,
            failed: 0,
            error: Some("cloud sync requires a paid plan".into()),
            plan_required: true,
        },
        Err(e) => sync_client::mark_all_failed(&store, &pending_items, &e.to_string())?,
    };
    drop(db);

    Ok(SyncResult {
        synced_count: attempt.synced as i64,
        failed_count: attempt.failed as i64,
        total_count,
        plan_required: attempt.plan_required,
    })
}

/// Delete a processed offline queue item (scoped).
#[tauri::command]
pub async fn delete_offline_item_scoped(
    id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    validate_not_empty("id", &id).map_err(|e| AppError::Invalid(e.to_string()))?;

    let (session, conn) = state.resolve_scope(&session_token)?;
    // F-017: enforce per-domain permission on this scoped command.
    require_permission_for_session(&state, &session, permissions::SYNC_MANAGE).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    store.delete_offline_item(&id)?;
    drop(db);

    tracing::info!(id, "offline queue item deleted (scoped)");
    Ok(())
}

/// Requeue a dead-lettered remote item (scoped).
#[tauri::command]
pub async fn requeue_remote_failure_scoped(
    args: RequeueRemoteFailureArgs,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    validate_not_empty("itemId", &args.item_id).map_err(|e| AppError::Invalid(e.to_string()))?;

    let (session, conn) = state.resolve_scope(&session_token)?;
    // F-017: enforce per-domain permission on this scoped command.
    require_permission_for_session(&state, &session, permissions::SYNC_MANAGE).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    run_requeue_remote_failure(&db, &args.item_id)?;
    drop(db);

    tracing::info!(item_id = %args.item_id, "dead-lettered remote item requeued (scoped)");
    Ok(())
}

/// List retained remote-application failures (scoped).
#[tauri::command]
pub async fn list_remote_failures_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<RemoteSyncFailureDto>, AppError> {
    let (_session, conn) = state.resolve_scope(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let failures = run_list_remote_failures(&db)?;
    drop(db);
    Ok(failures)
}
